### PR TITLE
Remove boilerplate code from implementations

### PR DESCRIPTION
**Pull request by imbryk**
The duplicated code from ViewModelBinder implementations have been
moved to the base class.
To allow this, a `Model` generic had to be changed to `BaseObservable`
and fields setters have too call `notifyPropertyChanged`

### CODE REVIEW

File: app/src/main/java/com/doctoror/menudatabinding/mvvm/ViewModelBinder.kt [hunk 1]
```diff
@@ -15,59 +15,73 @@
...
     var target: Target? = null
         set(value) {
             field = value
-            if (value != null) {
-                val model = model
-                if (model != null) {
-                    bind(value, model)
-                }
-            }
+            attemptBind(value, model)
```
Comment (imbryk) [inline app/src/main/java/com/doctoror/menudatabinding/mvvm/ViewModelBinder.kt]: this simplifies the above code a bit - the fields are implicitly made final by passing to the method, and there we can make a simple nullcheck

File: app/src/main/java/com/doctoror/menudatabinding/mvvm/ViewModelBinder.kt [hunk 2]
```diff
@@ -15,59 +15,73 @@
...
         }
 
     var model: Model? = null
         set(value) {
             val oldValue = field
-            field = value
-            if (value == null && oldValue != null) {
```
Comment (imbryk) [inline app/src/main/java/com/doctoror/menudatabinding/mvvm/ViewModelBinder.kt]: clear must be called even if value is not set to null - i.e. if a new model is set, old should stop being observed

File: app/src/main/java/com/doctoror/menudatabinding/mvvm/ViewModelBinder.kt [hunk 3]
```diff
@@ -15,59 +15,73 @@
...
     /**
      * Invoked when the [Model] is cleared.
      * This is a good place to remove OnPropertyChangeCallbacks.
      *
-     * @param oldValue the previous [Model], if was set.
+     * @param oldModel the previous [Model], if was set.
      */
-    internal open fun onModelCleared(oldValue: Model) {
-        // May be implemented by subclasses
+    private fun onModelCleared(oldModel: Model) {
```
Comment (imbryk) [inline app/src/main/java/com/doctoror/menudatabinding/mvvm/ViewModelBinder.kt]: now this (and the `onModelSet` method) is the major proposed change. The one thing I really disliked in that the implementations would have to create the observers every time. With this approach the binder implementations is much leaner

File: app/src/main/java/com/doctoror/menudatabinding/mvvm/ViewModelBinder.kt [hunk 4]
```diff
@@ -15,59 +15,73 @@
...
      */
-    internal abstract fun bind(target: Target, model: Model)
+    abstract fun bind(target: Target, model: Model)
+
+    abstract fun onPropertyChanged(model: Model, propertyId: Int)
+
+    private val modelChangeCallback =
+            object : Observable.OnPropertyChangedCallback() {
+                override fun onPropertyChanged(o: Observable, propertyId: Int) {
+                    @Suppress("UNCHECKED_CAST")
```
Comment (imbryk) [inline app/src/main/java/com/doctoror/menudatabinding/mvvm/ViewModelBinder.kt]: It is a pity, but we cannot do a proper `is` check with a generic... on the other hand, the subscription is done within a private method, so we now exactly, that we have subscribed to `model`, and `observable` passed in this method **is** of `Model` class

File: app/src/main/java/com/doctoror/menudatabinding/demo/mvvm/DemoToolbarBinder.kt [hunk 5]
```diff
@@ -15,37 +15,18 @@
  */
 package com.doctoror.menudatabinding.demo.mvvm
 
-import android.databinding.Observable
 import android.widget.Toolbar
+import com.doctoror.menudatabinding.BR
 import com.doctoror.menudatabinding.demo.DemoViewModel
 import com.doctoror.menudatabinding.mvvm.ViewModelBinder
 
 class DemoToolbarBinder : ViewModelBinder<Toolbar, DemoViewModel>() {
```
Comment (imbryk) [inline app/src/main/java/com/doctoror/menudatabinding/demo/mvvm/DemoToolbarBinder.kt]: As you can see - a lot of boilerplate removed here

File: app/src/main/java/com/doctoror/menudatabinding/demo/mvvm/DemoMenuBinder.kt [hunk 6]
```diff
@@ -15,39 +15,30 @@
...
 
-import android.databinding.Observable
 import android.view.Menu
+import android.view.MenuItem
+import com.doctoror.menudatabinding.BR
 import com.doctoror.menudatabinding.R
 import com.doctoror.menudatabinding.demo.DemoViewModel
 import com.doctoror.menudatabinding.mvvm.ViewModelBinder
 
 class DemoMenuBinder : ViewModelBinder<Menu, DemoViewModel>() {
```
Comment (imbryk) [inline app/src/main/java/com/doctoror/menudatabinding/demo/mvvm/DemoMenuBinder.kt]: same here - a lot of boilerplate removed

File: app/src/main/java/com/doctoror/menudatabinding/demo/DemoViewModel.kt [hunk 7]
```diff
@@ -15,11 +15,25 @@
...
 
     val mainContentText = "Hello World"
-    val toolbarMoved = ObservableBoolean(false)
-    val moveToolbarOptionEnabled = ObservableBoolean(true)
+
+    @Bindable
+    var toolbarMoved = false
+        set(value) {
+            field = value
+            notifyPropertyChanged(BR.toolbarMoved)
```
Comment (imbryk) [inline app/src/main/java/com/doctoror/menudatabinding/demo/DemoViewModel.kt]: This is the biggest con of proposed change - to make things work we must manually notify about property change. On the other hand I prefer to do this in the model once, and have a lean `Binder` which can be easily changed

Comment (Doctoror) [reply app/src/main/java/com/doctoror/menudatabinding/demo/DemoViewModel.kt]: Very good.

File: app/src/main/java/com/doctoror/menudatabinding/demo/DemoViewModel.kt [hunk 8]
```diff
@@ -15,11 +15,25 @@
...
+        }
+
+    @Bindable
+    var moveToolbarOptionEnabled = false
```
Comment (Doctoror) [inline app/src/main/java/com/doctoror/menudatabinding/demo/DemoViewModel.kt]: Was `true` by default in this model. And should be.

File: app/src/main/java/com/doctoror/menudatabinding/demo/mvvm/DemoMenuBinder.kt [hunk 9]
```diff
@@ -15,39 +15,30 @@
...
+    override fun bind(target: Menu, model: DemoViewModel) {
+        item = target.findItem(R.id.actionMoveToolbar)
+        update(model)
     }
 
-    override fun onModelCleared(oldValue: DemoViewModel) {
-        super.onModelCleared(oldValue)
-        oldValue.moveToolbarOptionEnabled.removeOnPropertyChangedCallback(
-                itemEnabledPropertyChangeCallback)
+    private fun update(model: DemoViewModel) {
```
Comment (Doctoror) [inline app/src/main/java/com/doctoror/menudatabinding/demo/mvvm/DemoMenuBinder.kt]: Would rather call this `updateMoveToolbarItem`.

File: app/src/main/java/com/doctoror/menudatabinding/demo/mvvm/DemoToolbarBinder.kt [hunk 10]
```diff
@@ -15,37 +15,18 @@
...
-    }
-
-    override fun onModelCleared(oldValue: DemoViewModel) {
-        super.onModelCleared(oldValue)
-        oldValue.toolbarMoved.removeOnPropertyChangedCallback(
-                toolbarHiddenOnPropertyChangeCallback)
-    }
-
     override fun bind(target: Toolbar, model: DemoViewModel) {
-        if (model.toolbarMoved.get()) {
```
Comment (Doctoror) [inline app/src/main/java/com/doctoror/menudatabinding/demo/mvvm/DemoToolbarBinder.kt]: You may have confused this with copypasta, but this block actually binded to the first-time state with no animations, while the other part (the `OnPropertyChanged` handler) actually animated it.

File: app/src/main/java/com/doctoror/menudatabinding/mvvm/ViewModelBinder.kt [hunk 11]
```diff
@@ -15,59 +15,73 @@
...
      * This is a good place to remove OnPropertyChangeCallbacks.
      *
-     * @param oldValue the previous [Model], if was set.
+     * @param oldModel the previous [Model], if was set.
      */
-    internal open fun onModelCleared(oldValue: Model) {
-        // May be implemented by subclasses
+    private fun onModelCleared(oldModel: Model) {
+        oldModel.removeOnPropertyChangedCallback(
+                modelChangeCallback)
```
Comment (Doctoror) [inline app/src/main/java/com/doctoror/menudatabinding/mvvm/ViewModelBinder.kt]: Can fit into a single line :)

File: app/src/main/java/com/doctoror/menudatabinding/mvvm/ViewModelBinder.kt [hunk 12]
```diff
@@ -15,59 +15,73 @@
...
      * Invoked when the [Model] is set.
      * This is a good place to add OnPropertyChangeCallbacks.
-     * @param model the new [Model], that is being set.
+     * @param newModel the new [Model], that is being set.
      */
-    internal open fun onModelSet(model: Model) {
-        // May be implemented by subclasses
+    private fun onModelSet(newModel: Model) {
+        newModel.addOnPropertyChangedCallback(
+                modelChangeCallback)
```
Comment (Doctoror) [inline app/src/main/java/com/doctoror/menudatabinding/mvvm/ViewModelBinder.kt]: Can fit into a single line :)

File: app/src/main/java/com/doctoror/menudatabinding/mvvm/ViewModelBinder.kt [hunk 13]
```diff
@@ -15,59 +15,73 @@
...
+    }
+
+    private fun attemptBind(target: Target?, model: Model?) {
+        if (target != null && model != null) {
```
Comment (Doctoror) [inline app/src/main/java/com/doctoror/menudatabinding/mvvm/ViewModelBinder.kt]: A lot better than what I had :+1:

File: app/src/main/java/com/doctoror/menudatabinding/demo/mvvm/DemoToolbarBinder.kt (hunk 10 above)
Comment (imbryk) [reply app/src/main/java/com/doctoror/menudatabinding/demo/mvvm/DemoToolbarBinder.kt]: well - I did notice, but as a demo for PoC I wanted to see the BemoBinder as clear a possible ;)

**Pull request merged**